File: YoutubeQueuer/YoutubeQueuer.Lib/Services/YoutubePlaylistsService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Google;
using Google.Apis.Auth.OAuth2;
using Google.Apis.YouTube.v3;
using Google.Apis.YouTube.v3.Data;
using YoutubeQueuer.Common;
using YoutubeQueuer.Lib.Models;
using YoutubeQueuer.Lib.Providers.Abstract;
using YoutubeQueuer.Lib.Services.Abstract;

namespace YoutubeQueuer.Lib.Services
{
    internal class YoutubePlaylistsService : IYoutubePlaylistsService
    {
        private readonly IYoutubeServiceProvider _youtubeServiceProvider;
        private readonly IYoutubeConstsProvider _youtubeConstsProvider;

        public YoutubePlaylistsService(IYoutubeServiceProvider youtubeServiceProvider,
            IYoutubeConstsProvider youtubeConstsProvider)
        {
            _youtubeServiceProvider = youtubeServiceProvider;
            _youtubeConstsProvider = youtubeConstsProvider;
        }

        public IEnumerable<YoutubePlaylistModel> GetUserPlaylists(UserCredential credential)
        {
            var youtube = _youtubeServiceProvider.GetYoutubeService(credential);

            var request = youtube.Playlists.List(_youtubeConstsProvider.PlaylistsListParts);

            request.MaxResults = 50;
            request.Mine = true;

            var result = request.Execute();

            var chann = youtube.Channels.List("id,snippet,contentDetails");
            chann.Mine = true;

            var playlists = MapPlaylists(result);

            return playlists;
        }

        private static IEnumerable<YoutubePlaylistModel> MapPlaylists(PlaylistListResponse result)
        {
            var playlists = result.Items.Select(x => new YoutubePlaylistModel
            {
                Name = x.Snippet.Title,
                Id = x.Id
            }).ToList();

            playlists.Add(new YoutubePlaylistModel
            {
                Id = "WL",
                Name = "Watch later"
            });
            return playlists;
        }

        public Result AddVideosToPlaylist(IEnumerable<string> videoIds, string playlistId, UserCredential credential)
        {
            try
            {
                var youtube = _youtubeServiceProvider.GetYoutubeService(credential);

                var items = videoIds
                    .Where(x => !string.IsNullOrWhiteSpace(x))
                    .Select(x => ToPlaylistItem(playlistId, x)).ToList();

                foreach (var playlistItem in items)
                {
                    AddVideoToPlaylist(youtube, playlistItem);
                }
                return Result.Succeed();
            }
            catch (Exception e)
            {
                return Result.Fail();
            }
        }

        private void AddVideoToPlaylist(YouTubeService youtube, PlaylistItem playlistItem)
        {
            try
            {
                var req = youtube.PlaylistItems.Insert(playlistItem, _youtubeConstsProvider.InsertPlaylistItemsParts);
                req.Execute();
            }
            catch (GoogleApiException e)
            {
                if (e.Error.Code != 409)
                {
                    throw;
                }
            }
        }

        private PlaylistItem ToPlaylistItem(string playlistId, string videoId)
        {
            return new PlaylistItem
            {
                Id = playlistId,
                Snippet = new PlaylistItemSnippet
                {
                    PlaylistId = playlistId,
                    ResourceId = new ResourceId
                    {
                        VideoId = videoId,
                        Kind = _youtubeConstsProvider.VideoResourceKind,
                        PlaylistId = playlistId
                    }
                },
                Kind = _youtubeConstsProvider.YoutubePlaylistItemKind
            };
        }
    }
}

File: YoutubeQueuerJS/server/services/youtube-playlists-service.js
var google = require('googleapis');
var googleAuth = require('./../google-auth');
var youtubeServiceProvider = require('./../youtube-service-provider');
var feedSchedulingSerivce = require('./feed-scheduling-service');

var playlistsService = {};

function toPlaylistModel(basePlaylist) {
    return {
        playlistId: basePlaylist.id,
        name: basePlaylist.snippet.title,
        scheduledTime: null
    };
}

playlistsService.getPlaylists = function() {
    var youtube = youtubeServiceProvider.getYoutubeService();

    var requestParams = {
        part: 'id,snippet',
        mine: true,
        maxResults: 50,
    };

    return new Promise((resolve, reject) => {
        var playlists = [];
        youtube.playlists.list(requestParams,
            (error, data, response) => {
                if (error) {
                    console.error('Error while getting playlists:' + error);
                    reject(error);
                }
                if (data) {
                    var mapped = data.items.map(toPlaylistModel);
                    mapped.unshift({
                        playlistId: "WL",
                        name: "Watch later"
                    });
                    playlists = mapped;
                    feedSchedulingSerivce.getScheduledTimesForPlaylists()
                        .then(schedulingSettings => {
                            for (let playlist of playlists) {
                                var setting = schedulingSettings.find(s => s.playlistId === playlist.playlistId);
                                if (setting) {
                                    playlist.scheduledTime = setting.scheduledTime;
                                }
                            }
                            resolve(playlists);
                        }).catch(() => resolve(playlists));
                }
            });
    });
};

playlistsService.addVideosToPlaylist = function(videoIds, playlistId) {
    'use strict';
    var youtube = youtubeServiceProvider.getYoutubeService();

    var videosToAdd = videoIds.map(v => {
        return {
            resource: {
                snippet: {
                    playlistId: playlistId,
                    resourceId: {
                        'videoId': v,
                        'kind': 'youtube#video',
                    }
                }
            },
            part: 'id,snippet',
            resourceId: {
                'videoId': v,
                'kind': 'youtube#video',
            }
        }
    });

    for (let videoPayload of videosToAdd) {
        try {
            youtube.playlistItems.insert(videoPayload,
                function(error, data) {
                    if (error) {
                        if (error.code !== 409) {
                            console.log('error while adding videos to playlist:', error);
                        } else {
                            console.log('Video already in playlist');
                        }
                    }
                });
        }
        catch (error) {
            console.log(error);
        }
    }
}

module.exports = playlistsService;